### ISSUE TITLE
Installation of overrides stops after a function that already is overridden is found

### ISSUE THREAD

**Issue by musicpanda** [thirtybees/thirtybees#1507]
When installing a theme with 8 override files I got an error message that one override could not be installed because that method in that file was already overridden. When I checked only one override had been installed. Obviously the installation had stopped after the conflict had been found. Expected behavior would be that it would try to install all eight and report whcih of them caused conflicts.

See also my post on the forum: https://forum.thirtybees.com/topic/5777-panda-question/

**Comment by getdatakick** [comment]
The expected behaviour should be *all or nothing*. On failure, no new overrides should be present, and module should not be marked as installed. 

I agree that the system should attempt to install all overrides and report all conflicts.

**Comment by musicpanda** [comment]
"On failure, no new overrides should be present, and module should not be marked as installed."

It is very problematic to mark the blocking presence of an existing override as a failure. It makes it de facto impossible to install a module for which such a situation exists. When you manually merge the overrides the situation that there is already an override will remain.

**Comment by getdatakick** [comment]
That's true. 

I guess that we could implement some interactive 'Resolve conflicts' page during module installation, where people would merge the two overrides. Something like resolving conflicts when merging in git.

**Comment by musicpanda** [comment]
That sounds to me like overengineering. 

People like to do editing in their own editor instead of some window. In addition you will need some kind of error handling. People will make errors when adding the code. That can as simple as forgetting a semicolon.

So my preference is trusting that the developer will merge the code.

If you want something extra my preference would be to take file date of the override file and annoy the backoffice with some warning as long it hasn't been changed.

**Comment by getdatakick** [comment]
This is a catch-22 situation. 

- we don't want to have a situation when we only partially install the module. It is very likely that the store will not work correctly after. 
- by blocking module installation because some override already exists, it makes it impossible to install the module. The only solution would be to modify the module before installation, remove override, and merge it manually. Then the module installation would work.

My proposed solution solves this problem (partially), as it gives option to merge the override during installation. Merchant can decide to abort the module installation, or attempt to make a merge.  

There could also be a button that would say 'keep current override but proceed with installation'. This would result in incomplete/partial installation of a module. We don't want this, but at least it would be decision of a person that is installing the module. 

Note that there is also significant problem with uninstallation of module for which we have made override merge.

**Comment by musicpanda** [comment]
Yes, overrides have problems. That is exactly why Prestashop is gradually dropping them.

But despite such "problems" Prestashop managed to grow to one of the biggest webshop platforms. I think that is no coincidence. The strength of the old Prestashop is that the code is simple and straightforward and bugs are usually easy to locate and solve. Sure, this appeals to a certain type of people and it doesn't work for everyone. But the tribe for who it works is big enough.

Prestashop's 1.7 development destroyed this simplicity and that is why people chose to fork it and start Thirty Bees.

For my taste with the complicated solutions you are offering for conflicting modules you are going the road of Prestashop 1.7 and destroy so the soul of Thirty Bees. Code is becoming less transparent and more and more bugs happen not in the functional code but in the layers around it that are supposed to prevent problems.
